network/transport/grpc/connection.go

```diff
@@ -169,62 +147,105 @@ func (mc *conn) verifyOrSetPeerID(id transport.PeerID) bool {
-		return fmt.Errorf("peer is already connected (method=%s)", method)
-	}
-	if _, exists := mc.grpcOutboundStreams[method]; exists {
-		return fmt.Errorf("peer is already connected (method=%s)", method)
+	methodName := protocol.MethodName()
+	if mc.streams[methodName] != nil {
+		return false
 	}
 
-	mc.grpcInboundStreams[method] = serverStream
-
-	// when the gRPC stream cancels, cancel the connection's context
 	if mc.ctx == nil {
 		mc.ctx, mc.cancelCtx = context.WithCancel(context.Background())
 	}
+
+	mc.streams[methodName] = stream
+	mc.outboxes[methodName] = make(chan interface{}, 20)
+
+	mc.startReceiving(protocol, stream)
+	mc.startSending(protocol, stream)
+
+	// A connection can have multiple active streams, but if one of them is closed, all of them should be closed, also closing the underlying connection.
 	go func(cancel func()) {
-		// Wait for the serverStream to close. Then remove it, and if it was the last one, callback
-		<-serverStream.Context().Done()
+		<-stream.Context().Done()
 		cancel()
-		mc.mux.Lock()
-		defer mc.mux.Unlock()
-		// Remove this stream from the inbound stream registry
-		delete(mc.grpcInboundStreams, method)
-		// If empty, remove.
-		if len(mc.grpcInboundStreams) == 0 {
-			mc.inboundStreamsClosedCallback(mc)
-		}
 	}(mc.cancelCtx)
-	return nil
+
+	return true
 }
 
-func (mc *conn) open(tlsConfig *tls.Config, connectedCallback func(grpcConn *grpc.ClientConn) bool) {
+func (mc *conn) startReceiving(protocol Protocol, stream Stream) {
+	go func() {
+		for {
+			message := protocol.CreateEnvelope()
+			err := stream.RecvMsg(message)
+			if err != nil {
+				errStatus, isStatusError := status.FromError(err)
+				if isStatusError && errStatus.Code() == codes.Canceled {
+					log.Logger().Infof("%s: Peer closed connection (peer=%s)", protocol.MethodName(), mc.Peer())
+				} else {
+					log.Logger().Warnf("%s: Peer connection error (peer=%s): %v", protocol.MethodName(), mc.Peer(), err)
+				}
+				mc.mux.Lock()
+				mc.cancelCtx()
+				mc.mux.Unlock()
+				break
+			}
+
+			err = protocol.Handle(mc.Peer(), message)
+			if err != nil {
+				log.Logger().Warnf("%s: Error handling message %T (peer=%s): %v", protocol.MethodName(), protocol.UnwrapMessage(message), mc.Peer(), err)
+			}
+		}
+	}()
+}
+
+func (mc *conn) startSending(protocol Protocol, stream Stream) {
+	outbox := mc.outboxes[protocol.MethodName()]
+	done := mc.ctx.Done()
+
+	go func() {
+		for {
+			select {
+			case _ = <-done:
+				goto closeClientStream
+			case envelope := <-outbox:
+				err := stream.SendMsg(envelope)
+				if err != nil {
+					log.Logger().Warnf("Unable to send message %T, message is dropped (peer=%s): %v", envelope, mc.Peer(), err)
+				}
+			}
+		}
+	closeClientStream:
+		clientStream, ok := stream.(grpc.ClientStream)
+		if ok {
+			err := clientStream.CloseSend()
+			if err != nil {
+				log.Logger().Warnf("Error while closing client for gRPC stream %s: %v", protocol.MethodName(), err)
```

Comment:
Not sure when this can happen. Could be already closed, or the peer doesn't accept the close message. I figured it's useful to log to find out (otherwise I'd have just ignored it), but probably not important/severe enough for error log?